Introduce WTF::unsignedCast() to cast a signed type variable into an unsigned type one

cdumez opened the pull request:
#### 71d38ea602290ddb3bf462074ab971898083a14c
<pre>
Introduce WTF::unsignedCast() to cast a signed type variable into an unsigned type one
<a href="https://bugs.webkit.org/show_bug.cgi?id=292702">https://bugs.webkit.org/show_bug.cgi?id=292702</a>

Reviewed by Darin Adler.

Introduce WTF::unsignedCast() to cast a signed type variable into an unsigned type one.
It relies on std::make_unsigned_t internally to figure out the unsigned type to cast to.

Do some adoption in the codebase.

* Source/JavaScriptCore/b3/B3Common.h:
(JSC::B3::chillUDiv):
(JSC::B3::chillUMod):
(JSC::B3::rotateRight):
(JSC::B3::rotateLeft):
* Source/JavaScriptCore/b3/B3ComputeDivisionMagic.h:
(JSC::B3::computeDivisionMagic):
* Source/JavaScriptCore/b3/B3ReduceStrength.cpp:
* Source/JavaScriptCore/b3/testb3.h:
(modelLoad):
* Source/JavaScriptCore/b3/testb3_5.cpp:
(modelCompare):
* Source/WTF/wtf/CheckedArithmetic.h:
(WTF::safeCast):
* Source/WTF/wtf/EnumTraits.h:
(WTF::enumNamesSize):
* Source/WTF/wtf/Hasher.h:
(WTF::requires):
* Source/WTF/wtf/HexNumber.h:
(WTF::Internal::appendHex):
* Source/WTF/wtf/LEBDecoder.h:
(WTF::LEBDecoder::decodeInt):
* Source/WTF/wtf/MathExtras.h:
(WTF::clzConstexpr):
(WTF::clz):
(WTF::ctzConstexpr):
(WTF::ctz):
(WTF::negate):
* Source/WTF/wtf/StdLibExtras.h:
(WTF::requires):
(WTF::safeCast): Deleted.
* Source/WTF/wtf/text/IntegerToStringConversion.h:
(WTF::numberToStringSigned):
(WTF::writeIntegerToBuffer):
(WTF::lengthOfIntegerAsString):
* Source/WTF/wtf/text/StringBuffer.h:
* Source/WTF/wtf/text/StringCommon.h:
(WTF::isLatin1):
* Source/WTF/wtf/text/StringHasher.h:
(WTF::StringHasher::DefaultConverter::convert):
* Source/WebCore/css/parser/CSSParser.cpp:
(WebCore::CSSParser::consumeFontFeatureValuesRuleBlock):
* Source/WebCore/platform/graphics/GlyphBuffer.h:
(WebCore::GlyphBuffer::checkedStringOffsetAt const):
* Source/WebGPU/WGSL/AttributeValidator.cpp:
(WGSL::AttributeValidator::visit):
* Source/WebKit/Scripts/webkit/tests/SerializedTypeInfo.cpp:
(WebKit::enumValueForIPCTestAPI):

Canonical link: <a href="https://commits.webkit.org/294691@main">https://commits.webkit.org/294691@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/524723b5d6412b004b443029b0949bc865a17995

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/102657 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/22331 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/12648 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/107822 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/53298 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/104696 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/22638 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/30832 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/78086 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/35059 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/105663 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/17544 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/92655 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/58418 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/102185 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/17384 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/10685 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/52655 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/95332 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/87205 "Passed tests") | [❌ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/10756 "Found 1 new test failure: http/tests/workers/service/registerServiceWorkerClient-after-network-process-crash.html (failure)") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/110198 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/101268 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/29794 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/21971 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/87076 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/30158 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/88846 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/86681 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/22074 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/31497 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/9233 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/24047 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/29721 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/35038 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/124900 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/29529 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/34671 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/32855 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/31091 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rniwa commented on Source/JavaScriptCore/b3/B3Common.h:
```diff
@@ -72,9 +73,8 @@ static IntType chillMod(IntType numerator, IntType denominator)
-    typedef typename std::make_unsigned<IntType>::type UnsignedIntType;
-    UnsignedIntType unsignedNumerator = static_cast<UnsignedIntType>(numerator);
-    UnsignedIntType unsignedDenominator = static_cast<UnsignedIntType>(denominator);
+    auto unsignedNumerator = unsignedCast(numerator);
```
Isn't this more of `castToUnsigned` or `castUnsigned` than `unsignedCast`? It's not like cast operation is unsigned.

cdumez replied:
We have `byteCast`, `spanConstCast`, `std::bit_cast` already so it's not that different. Let's see what @darinadler thinks since he suggested the naming in this PR.

rniwa replied:
I see.

darinadler replied:
I agree with both of you. It families better with other casts as `unsigned_cast` or `unsignedCast` and I can imagine giving it a clearer name.

cdumez replied:
Maybe `castToUnsigned` is clearer then.

rniwa replied:
FWIW, `unsignedCast` is fine to me given the precedence of other functions.

darinadler commented on Source/JavaScriptCore/b3/testb3.h:
```diff
@@ -555,7 +555,7 @@ EffectiveType modelLoad(EffectiveType value)
     u.original = value;
     if (std::is_signed<LoadedType>::value)
         return static_cast<EffectiveType>(u.loaded);
```
Can we just remove these two lines of code? I think they aren't needed to handle the signed case correctly and efficiently.

darinadler commented on Source/WTF/wtf/StdLibExtras.h:
```diff
@@ -1250,6 +1242,11 @@ template<ByteType T, typename U> constexpr auto byteCast(const U& value)
     return ByteCastTraits<U>::template cast<T>(value);
 }
 
+template<typename T> constexpr auto unsignedCast(T value) requires (std::is_arithmetic_v<T>)
```
We want a compile error if this is used on a floating point type. So maybe `is_integral` is better than `is_arithmetic`. But ideally this would also work on anything else that `make_unsigned` works on, like character types and enumerations. So maybe no `requires` clause at all is best, and when the type is one that make_unsigned doesn't work with such as a floating point number or a non-scalar the body of the function will fail to compile since it's ill-formed?

darinadler commented on Source/WTF/wtf/text/StringCommon.h:
```diff
@@ -117,7 +117,7 @@ size_t strlenSpan(std::span<T, Extent> span) requires(sizeof(T) == 1)
 {
-    using UnsignedCharacterType = typename std::make_unsigned<CharacterType>::type;
+    using UnsignedCharacterType = std::make_unsigned_t<CharacterType>;
     return static_cast<UnsignedCharacterType>(character) <= static_cast<UnsignedCharacterType>(0xFF);
```
```suggestion
    return unsignedCast(character) <= 0xFFu;
```

darinadler commented on Source/WebKit/Scripts/webkit/tests/SerializedTypeInfo.cpp:
```diff
@@ -115,7 +119,7 @@ namespace WebKit {
 template<typename E> uint64_t enumValueForIPCTestAPI(E e)
 {
-    return static_cast<std::make_unsigned_t<std::underlying_type_t<E>>>(e);
+    return unsignedCast(enumToUnderlyingType(e));
```
I think if we removed the `requires` from the implementation of `unsignedCast` we could do this with `unsignedCast` alone because `make_unsigned` handles this case.
```suggestion
    return unsignedCast(e);
```

darinadler commented on Source/WebKit/Scripts/webkit/tests/SerializedTypeInfo.cpp:
```diff
@@ -71,14 +71,18 @@
 #include <WebCore/ScrollingStateFrameHostingNodeWithStuffAfterTuple.h>
 #include <WebCore/TimingFunction.h>
+#include <wtf/CreateUsingClass.h>
+#include <wtf/EnumTraits.h>
```
I don't think we need this one.

cdumez committed:
Introduce WTF::unsignedCast() to cast a signed type variable into an unsigned type one
https://bugs.webkit.org/show_bug.cgi?id=292702

Reviewed by Darin Adler.

Introduce WTF::unsignedCast() to cast a signed type variable into an unsigned type one.
It relies on std::make_unsigned_t internally to figure out the unsigned type to cast to.

Do some adoption in the codebase.

* Source/JavaScriptCore/b3/B3Common.h:
(JSC::B3::chillUDiv):
(JSC::B3::chillUMod):
(JSC::B3::rotateRight):
(JSC::B3::rotateLeft):
* Source/JavaScriptCore/b3/B3ComputeDivisionMagic.h:
(JSC::B3::computeDivisionMagic):
* Source/JavaScriptCore/b3/B3ReduceStrength.cpp:
* Source/JavaScriptCore/b3/testb3.h:
(modelLoad):
* Source/JavaScriptCore/b3/testb3_5.cpp:
(modelCompare):
* Source/WTF/wtf/CheckedArithmetic.h:
(WTF::safeCast):
* Source/WTF/wtf/EnumTraits.h:
(WTF::enumNamesSize):
* Source/WTF/wtf/Hasher.h:
(WTF::requires):
* Source/WTF/wtf/HexNumber.h:
(WTF::Internal::appendHex):
* Source/WTF/wtf/LEBDecoder.h:
(WTF::LEBDecoder::decodeInt):
* Source/WTF/wtf/MathExtras.h:
(WTF::clzConstexpr):
(WTF::clz):
(WTF::ctzConstexpr):
(WTF::ctz):
(WTF::negate):
* Source/WTF/wtf/StdLibExtras.h:
(WTF::requires):
(WTF::safeCast): Deleted.
* Source/WTF/wtf/text/IntegerToStringConversion.h:
(WTF::numberToStringSigned):
(WTF::writeIntegerToBuffer):
(WTF::lengthOfIntegerAsString):
* Source/WTF/wtf/text/StringBuffer.h:
* Source/WTF/wtf/text/StringCommon.h:
(WTF::isLatin1):
* Source/WTF/wtf/text/StringHasher.h:
(WTF::StringHasher::DefaultConverter::convert):
* Source/WebCore/css/parser/CSSParser.cpp:
(WebCore::CSSParser::consumeFontFeatureValuesRuleBlock):
* Source/WebCore/platform/graphics/GlyphBuffer.h:
(WebCore::GlyphBuffer::checkedStringOffsetAt const):
* Source/WebGPU/WGSL/AttributeValidator.cpp:
(WGSL::AttributeValidator::visit):
* Source/WebKit/Scripts/webkit/tests/SerializedTypeInfo.cpp:
(WebKit::enumValueForIPCTestAPI):

Canonical link: https://commits.webkit.org/294691@main
